rfcs/20200903-tensor-float-32-default.md

```diff
@@ -0,0 +1,46 @@
+# Enabling TensorFloat-32 by default
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [287](https://github.com/tensorflow/community/pull/287) |
+| **Author(s)** | Reed Wanderman-Milne (reedwm@google.com)             |
+| **Sponsor**   | Sanjoy Das (sanjoy@google.com)                       |
+| **Updated**   | 2020-09-03                                           |
+
+## Objective
+
+Enable [TensorFloat-32](https://blogs.nvidia.com/blog/2020/05/14/tensorfloat-32-precision-format/) by default.
+
+## Motivation
+
+TensorFloat-32, or TF32 for short, is a new math mode for Ampere GPUs. Support for TF32 was recently added to TensorFlow and [an RFC for the TF32 API](https://github.com/tensorflow/community/blob/master/rfcs/20200520-tensor-float-32.md) was accepted. This RFC proposes enabling TF32 by default.
+
+TF32 has the same numeric range as float32 but lower precision. The effect of enabling TF32 is that on Ampere GPUs, certain float32 ops, like matmul and convolutions, will be faster but use lower-precision.
+
+This RFC will serve as a place where we can collect feedback on potential cases where enabling TF32 will break users. We have enabled TF32 by default in the TensorFlow nightlies, but will disable it by TensorFlow 2.4’s release if this RFC is rejected.
+
+## User Benefit
+
+The benefit of turning on TF32 by default is that all Ampere users will get the performance benefit, even if they don’t know about TF32 and therefore do not call the API function to enable it.
+
+Turning on TF32 by default also makes the GPU behavior more consistent with TPUs. TPUs also use a lower-precision internal dtype (bfloat16) for matrix multiplications and convolutions by default when float32 is used.
+
+## Design Proposal
+
+The design is to flip the default value of [`tf.config.experimental.enable_tensor_float_32_execution`](https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_tensor_float_32_execution) from False to True.
```

Comment:
Would it be practical to provide a user control for setting the default type with a global toggle, for all numerical ops? That might be useful in case there isn't a one-size-fits-all.

Reply:
This is somewhat orthogonal to TensorFloat-32, since TF32 is not a TensorFlow dtype. Even with TF32 enabled, `tf.linalg.matmul` outputs float32 tensors when passed float32 inputs. The only effect of TF32 is increased performance and inputs to ops like matmul are rounded from 23 bits of precision to 10 bits of precision.

There should be an API to set the default dtype for ops like `tf.ones` that don't take an input tensor. But I think ops like matmul should continue to unconditionally output tensors of the same dtype as the inputs.